x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -1346,6 +1379,232 @@ private static Expression cast(org.elasticsearch.xpack.esql.core.expression.func
+         * Both the logical plan's children and the logical plan itself can be changed, so we cannot just to a replaceChild here.
+         */
+        private static LogicalPlan castInvalidMappedFieldInLogicalPlan(LogicalPlan plan, boolean addProject) {
+            List<? extends Expression> originalExpressions;
+            if (addProject) {
+                originalExpressions = plan.output();
+            } else {
+                originalExpressions = switch (plan) {
+                    case EsqlProject project -> project.projections();
+                    case OrderBy ob -> ob.order();
+                    case Aggregate agg -> agg.groupings();
+                    case MvExpand me -> List.of(me.target());
+                    case Eval eval -> eval.fields();
+                    case LookupJoin lj -> lj.config().leftFields();
+                    case Enrich enrich -> List.of(enrich.matchField());
+                    // The other types of plans are unexpected
+                    default -> throw new EsqlIllegalArgumentException("unexpected logical plan: " + plan);
+                };
+            }
+            Tuple<List<Alias>, List<Expression>> newChildren = castInvalidMappedFields(originalExpressions, true);
+            List<Alias> aliases = newChildren.v1();
+            List<Expression> newProjections = newChildren.v2();
+            if (aliases.isEmpty() == false) {
+                if (addProject) {
+                    return esqlProjectForInvalidMappedField(plan.source(), plan, aliases, newProjections);
+                }
+                switch (plan) {
+                    case EsqlProject p -> {
+                        return esqlProjectForInvalidMappedField(p.source(), p.child(), aliases, newProjections);
+                    }
+                    case OrderBy o -> {
+                        return new OrderBy(
+                            o.source(),
+                            o.child(),
+                            newProjections.stream().filter(e -> e instanceof Order).map(e -> (Order) e).collect(Collectors.toList())
+                        );
+                    }
+                    case Aggregate agg -> {
+                        // Both groupings and aggregates need to be updated, create new aggregates according to new groupings
+                        List<? extends NamedExpression> origAggs = agg.aggregates();
+                        List<NamedExpression> newAggs = new ArrayList<>(origAggs.size());
+                        for (int i = 0; i < origAggs.size() - newProjections.size(); i++) { // Add aggregate functions
+                            newAggs.add(origAggs.get(i));
+                        }
+                        for (Expression e : newProjections) { // Add groupings
+                            newAggs.add(Expressions.attribute(e));
+                        }
+                        return agg.with(evalForInvalidMappedField(agg.source(), agg.child(), aliases), newProjections, newAggs);
```

Comment:
Because we inject an `EVAL` here, there's a chance that we'll accidentally inject `EVAL`s multiple times. Which could lead to broken plans, when an upstream `EVAL` already replaced the multi-typed field attribute by a reference attribute of the same name. Or, put the other way around, since we're replacing a field by another one with the same name, downstream plans may have invalid references now.

This could happen in case that the reference resolution manages to resolve the whole query plan, with multiple references to multi-typed fields, before we start adding Evals.